src/equations/ideal_mhd_multiion_2d.jl

```diff
@@ -0,0 +1,888 @@
+            + 0.5 * vk2_plus_avg * mag_norm_avg - vk1_plus_avg * B2_avg * B1_avg - vk2_plus_avg * B2_avg * B2_avg - vk3_plus_avg * B2_avg * B3_avg   # Additional terms coming from the MHD non-conservative term (momentum eqs)
+            - B1_avg *  (vk2_minus_avg * B1_avg - vk1_minus_avg * B2_avg) - B3_avg * (vk2_minus_avg * B3_avg - vk3_minus_avg * B2_avg) )             # Terms coming from the non-conservative term 3 (induction equation!)
+
+      f = (f..., f1, f2, f3, f4, f5)
+    end
+  end
+
+  return SVector{nvariables(equations), real(equations)}(f)
+end
+
+"""
+# Calculate maximum wave speed for local Lax-Friedrichs-type dissipation
+  !!!ATTENTION: This routine is provisional. TODO: Update with the right max_abs_speed
+"""
+@inline function max_abs_speed_naive(u_ll, u_rr, orientation::Integer, equations::IdealMhdMultiIonEquations2D)
+  # Calculate fast magnetoacoustic wave speeds
+  # left
+  cf_ll = calc_fast_wavespeed(u_ll, orientation, equations)
+  # right
+  cf_rr = calc_fast_wavespeed(u_rr, orientation, equations)
+
+  # Calculate velocities
+  v_ll = zero(u_ll[1])
+  v_rr = zero(u_rr[1])
+  if orientation == 1
+    for k in eachcomponent(equations)
+      rho, rho_v1, _ = get_component(k, u_ll, equations)
+      v_ll = max(v_ll, abs(rho_v1 / rho))
+      rho, rho_v1, _ = get_component(k, u_rr, equations)
+      v_rr = max(v_rr, abs(rho_v1 / rho))
+    end
+  else #if orientation == 2
+    for k in eachcomponent(equations)
+      rho, rho_v1, rho_v2, _ = get_component(k, u_ll, equations)
+      v_ll = max(v_ll, abs(rho_v2 / rho))
+      rho, rho_v1, rho_v2, _ = get_component(k, u_rr, equations)
+      v_rr = max(v_rr, abs(rho_v2 / rho))
+    end
+  end
+
+  λ_max = max(abs(v_ll), abs(v_rr)) + max(cf_ll, cf_rr)
+end
+
+
+@inline function max_abs_speeds(u, equations::IdealMhdMultiIonEquations2D)
+  
+  v1 = zero(u[1])
+  v2 = zero(u[1])
+  for k in eachcomponent(equations)
+    rho, rho_v1, rho_v2, _ = get_component(k, u, equations)
+    v1 = max(v1, abs(rho_v1 / rho))
+    v2 = max(v2, abs(rho_v2 / rho))
+  end
+
+  cf_x_direction = calc_fast_wavespeed(u, 1, equations)
+  cf_y_direction = calc_fast_wavespeed(u, 2, equations)
+
+  return (abs(v1) + cf_x_direction, abs(v2) + cf_y_direction)
+end
+
+
+"""
+Convert conservative variables to primitive
+"""
+function cons2prim(u, equations::IdealMhdMultiIonEquations2D)
+  @unpack gammas = equations
+  B1, B2, B3, _ = u
+
+  prim = (B1, B2, B3)
+  for k in eachcomponent(equations)
+    rho, rho_v1, rho_v2, rho_v3, rho_e = get_component(k, u, equations)
+    srho = 1 / rho
+    v1 = srho * rho_v1
+    v2 = srho * rho_v2
+    v3 = srho * rho_v3
+
+    p = (gammas[k] - 1) * (rho_e - 0.5 * (rho_v1 * v1 + rho_v2 * v2 + rho_v3 * v3
+                                 + B1 * B1 + B2 * B2 + B3 * B3))
+    prim = (prim..., rho, v1, v2, v3, p)
```

Comment:
Since this is very likely not type stable and `cons2prim` can be used in performance-critical situations, I strongly suggest to fix this here and elsewhere...